core/shared/src/main/scala/zio/LogAnnotation.scala

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2021 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+final class LogAnnotation[V] private (
```

Comment:
I know this is the same as in `zio-logging` and that `initial` and `combine` makes it possible to generalize for example the name stack and custom annotations but in practice I found that all the annotations I'm creating are such that `V` is an option type and it feels a bit annoying having to write `Option` and `Some` everywhere - would be nice if I want to add a log annotation of a request ID then I create a `LogAnnotation[RequestId]` without having to deal with specifying that there is by default no request ID, and that there can be only one request ID. 

Probably we could have this as an advanced log annotation and have something called `LogAnnotation` for the more simpler use case built on top of it?

Reply:
Yeah I tried to go in that direct by specifying default arguments for the `combine` and `render` parameters. It seems like the most idiomatic thing to do is just to specify some default value for `V` like `LogAnnotation("key", "default")`. That depends a little on your data type whether there is some default value you can use.

We could do something like:

```scala
def optional[V](identifier: String)(implicit tag: Tag[V]): LogAnnotation[V] =
  LogAnnotation(identifier, null.asInstanceOf[V])
```

That requires the implementers of `ZLogger` to deal with the possibility of null values, though it is a relatively low level interface. I'm also not sure the ergonomics are that much better because then you have to specify a type parameter like `LogAnnotation.optional[String]("key")`.

Reply:
One alternative idea:

- Let's have a different input and storage type (parameter) of log annotations. So we can have a `LogAnnotation[V, Option[V]]` that accepts a `V` when you annotate, but produces an `Option[V]` for the renderer.
- Maybe rename this type from `LogAnnotation` to something like `LogAnnotationBase` or `RawLogAnnotation`, I don't know; in order to make `LogAnnotation` a simplified version
- Make a simplified version that the users can find by default (so I would call this one `LogAnnotation`, probably) that defaults to the option-backed version. While more advanced annotations can still directly use the full interface to provide all types and combine/initial/render

Reply:
hmm.. why we need optional log annotations? Either I annotate a certain effect or I don't. Whatever is there gets logged, what is not, does not get logged?

Reply:
Or are we actually now mixing LogAnnotations with something like a `RequestContext`  ?

Reply:
If you look at the "built-in" annotations, `Name` for example is a stack of names, with default value being an empty list and `combine` adding the name to the stack, while `Throwable` is something that is either there or not, represented by an `Option[Throwable]` annotation.

Personally I'm using zio-logging's similar capability to annotate logs with contextual info like request id, user id, etc., I believe that is the primary reason for having these annotations user-defineable.

Reply:
yeah, I know the annotations in `zio-logging`.. I also annotate my logs with request-id, user-id, correlation-id etc.. but never had the need for a value anything else than Some(v), so the only reason for me to have this as a Option would be to say "by default this key should be there, but the value may or may not be there.. if its missing, something is wrong, please search for the error" .. I must admit, I only ever used the feature to work together with the `Slf4jLogger.makeWithAnnotationsAsMdc(...)` thingy which basically is a Map[String, String] behind the scenes. 

However, I see the beauty in having special types like `Option`s, `List`s, `Stack`s, etc., I'm just not sure we need to have "special" support for `Option`?

Reply:
I think we are on the same page, my idea was exactly to make this most common case more convenient. Maybe one difference that for me the optionality of these contextual annotations is quite common - for example I may log things about a request that I could not authenticate and assign with a user id

Reply:
I think the issue is if you use something like a `UUID` for your annotation. So in `zio-logging` as well as in this PR you would do something like `val correlationID = LogAnnotation("user-id", UUID.fromString("0-0-0-0-0")` (this is basically taken from the `zio-logging` documentation). 

This is fine but I think what @vigoo is saying is the `default` value here is somewhat arbitrary and in some cases there may not be a logical default value so it would be nice to not have to specify it at all, and write something like `val correlationId = LogAnnotation[UUID]("user-id")`

Reply:
> This is fine but I think what @vigoo is saying is the `default` value here is somewhat arbitrary and in some cases there may not be a logical default value so it would be nice to not have to specify it at all, and write something like `val correlationId = LogAnnotation[UUID]("user-id")`

Exactly. I would have something like 
```scala
object Annotations { 
 val UserId = LogAnnotation[String]("userId")
 val CorrelationId = LogAnnotation[CorrelationId]("correlationId")
}
```

and then in my controller/filter/whatever would do 

```scala
 ...
 import {Annotations => l}
 
 myEffect.logAnnotate(l.UserId, jwt.userId).logAnnotate(l.CorrelationId, c)
 ```
 
 So I guess the question is if we need to preserve the information that a annotation should be there, but was not filled (which we could model with optional values). If we skip optionals / empty list anyway when rendering, there's no need for that.

Reply:
on the other hand, if we have complex data types like 
```scala
val Actions = LogAnnotation[List]("actions", empty = List.empty, combine = (a,b) => a ++ b, render => v => Json.toObj(v) } 
```
we can't necessarly fix the render type to string.. e.g. in the above case, the list could be rendered as Json Object or Array or `v.mkString(",")` 

of course, a `ZLogger` could match on the return value of the `render` call and then e.g. decide what to do with a `Seq` ..

Reply:
At least from my perspective we want to keep the typed logging but just make it more convenient for the common case. Let me play with this.